ocaml/typing/layouts.ml

```diff
@@ -48,24 +48,36 @@ module Sort = struct
 
   let new_var () = Var (ref None)
 
-  let rec repr ~default : t -> t = function
+  let rec repr : t -> t = function
     | Const _ as t -> t
     | Var r as t -> begin match !r with
-      | None -> begin match default with
-        | None -> t
-        | Some const -> begin
-            let t = of_const const in
-            r := Some t;
-            t
-          end
-      end
+      | None -> t
       | Some s -> begin
-          let result = repr ~default s in
+          let result = repr s in
           r := Some result; (* path compression *)
```

Comment:
While we're in town: this `Some result` will allocate a new `Some` node, even when there's no path to compress. Might be better to have `if result != s then r := Some result`, but that introduces another branch. I think probably allocation is always worse than a straightforward branch, though.

Reply:
Yeah.  I preferred the simpler version because (a) it's not on the path we added calls to and are trying to optimize here (this is the non-defaulting repr, but we added calls to the defaulting one), and (b) the branch may not be well predicted.  But I guess the allocation is potentially long-lived, so probably the other way is better?  I'll change it.